std/process.d

```diff
@@ -1050,54 +1050,103 @@ private Pid spawnProcessPosix(scope const(char[])[] args,
+                                dirent* entry = readdir(dir);
+
+                                if (entry is null) break;
+                                if (entry.d_name[0] == '.') continue;
+
+                                int fd = atoi(cast(char*) entry.d_name);
+
+                                // Don't close stdin, stdout, stderr, or the directory file descriptor
+                                if (fd < lowfd || fd == opendirfd) continue;
+
+                                close(fd);
                             }
                         }
                         else
                         {
-                            // Fall back to closing everything.
-                            foreach (i; lowfd .. maxDescriptors)
+                            // This is going to allocate 8 bytes for each possible file descriptor from lowfd to r.rlim_cur
+                            if (maxDescriptors <= 128*1024)
                             {
-                                close(i);
+                                // NOTE: malloc() and getrlimit() are not on the POSIX async
+                                // signal safe functions list, but practically this should
+                                // not be a problem. Java VM and CPython also use malloc()
+                                // in its own implementation via opendir().
+                                import core.stdc.stdlib : malloc;
+                                import core.sys.posix.poll : pollfd, poll, POLLNVAL;
+
+                                immutable maxToClose = maxDescriptors - lowfd;
+
+                                // Call poll() to see which ones are actually open:
+                                auto pfds = cast(pollfd*) malloc(pollfd.sizeof * maxToClose);
+                                if (pfds is null)
+                                {
+                                    abortOnError(forkPipeOut, InternalError.malloc, .errno);
+                                }
+
+                                foreach (i; 0 .. maxToClose)
+                                {
+                                    pfds[i].fd = i + lowfd;
+                                    pfds[i].events = 0;
+                                    pfds[i].revents = 0;
+                                }
+
+                                if (poll(pfds, maxToClose, 0) < 0)
+                                    // couldn't use poll, use the slow path.
+                                    goto slowFDClose;
+
+                                foreach (i; 0 .. maxToClose)
+                                {
+                                    // POLLNVAL will be set if the file descriptor is invalid.
+                                    if (!(pfds[i].revents & POLLNVAL)) close(pfds[i].fd);
+                                }
+                            }
+                            else
+                            {
+slowFDClose:
```

Comment:
indent? also can we not use a `goto` here>

Reply:
We can, it's just better with a goto. The previous version had a boolean just to avoid a goto, which is goofy.

I don't know the requirements for labels, I always put them on the left completely. What indent to use?

Reply:
Fair. not sure about probes, but DMD uses 4 spaces less than the current indentation. so 
```
else
{
slowFDClose:
```
all indented as appropriate.

Reply:
This seems to agree, though maybe I should prefix with L https://github.com/dlang/phobos/blob/master/std/math/exponential.d#L3542

Reply:
ok, maybe it should be one indent less than the line labeled https://github.com/dlang/phobos/blob/master/std/string.d#L5819

Reply:
Updated

Reply:
Oof, for some reason github wasn't showing all your answers. We came to the same conclusion though...